### ISSUE TITLE
MVP Checkpointing

### ISSUE THREAD

**Issue by kasima** [omisego/research#36]

**Comment by smartcontracts** [comment]
# Checkpoints in Plasma MVP

## TL;DR

This document provides a brief overview of checkpoints in Plasma MVP. Unlike Plasma Cash, Plasma MVP supports "naive" checkpoints without changing any assumptions.

## Background

Plasma MVP is big. Without any checkpointing, Plasma MVP grows proportionally to the number of blocks and the number of transactions per block. The current implementation (65536 tx/block) grows at a rate of 68TB/yr. This is a huge amount of data, so we're motivated to devise mechanisms that reduce the total storage requirement.

## Checkpointing

Checkpointing is one way to reduce the storage requirement. We can implement basic checkpointing by having the operator submit checkpoint roots along with each block root. Checkpoint roots are calculated from the Merkle tree consisting of all outstanding UTXOs as of the submitted block. After 2 weeks, all transaction inputs must either come from this checkpoint block or some later block. 

The result of this system is that full nodes will only ever have to store 2 weeks worth of data. The total data storage requirement now becomes ~2.6 TB, ever. This is still large, but it's a much more reasonable amount of data to store and replicate. 

Note that it's possible for the operator to submit an invalid checkpoint root. Any invalid checkpoint root will force every use to exit, just as an invalid or withheld block. As a result, we haven't changed any Plasma MVP assumptions.

**Comment by pdobacz** [comment]
> all outstanding UTXOs as of the submitted block

How would that respect the deposits/exits that take place on the root chain? It feels like they might desynchronize the UTXO set (i.e. different observers might end up with a slightly different one, while they need to validate against the same merkle root). It looks like there's a rule to handle these required.
